File: Stack/Stack_by_linked_list/stack.cpp
#include <iostream>
using namespace std; 

class StackNode{
    public:
        int data ; 
        StackNode* next ; 
}; 

StackNode* newNode(int data){
    StackNode* node = new StackNode();
    node->data = data ; 
    node->next = NULL ; 

    return node ; 
}

int isEmpty(StackNode* node){
    return !node;
}

void push(StackNode** node, int data){
    StackNode* stacknode = newNode(data) ; 
    stacknode->next = *node ; 
    *node = stacknode ; 
    cout<<data<<"pushed to stack \n" ;  
}

int pop(StackNode** root){
    if(isEmpty(*root)){
        return 0 ; 
    }
    StackNode* temp = *root ; 
    *root = (*root)->next ; 
    int popped = temp->data ;
    free(temp) ; 

    return popped ; 
}

int peek(StackNode* root){
    if(isEmpty(root)){
        return 0 ; 
    }
    return root->data ; 
}

int main()
{
    StackNode* root = NULL;
  
    push(&root, 10);
    push(&root, 20);
    push(&root, 30);
  
    cout << pop(&root) << " popped from stack\n";
  
    cout << "Top element is " << peek(root) << endl;
      
    cout<<"Elements present in stack : ";
     //print all elements in stack :
    while(!isEmpty(root))
    {
        // print top element in stack
        cout<<peek(root)<<" ";
        // remove top element in stack
        pop(&root);
    }
  
    return 0;
}

File: Stack/Stack_by_linked_list/stack.js
class StackNode{
    constructor(data){
        this.data = data ; 
        this.next = null ; 
    }
}

var root = null ;
function isEmpty(){
    return root == null ; 
} 
function push(data){
    var newstack = new StackNode(data) ;
    if(isEmpty()){
        root = newstack ;  
    }else{
        var temp = root ;  
        newstack.next = temp ; 
        root = newstack ;  
    }
}

function pop(){
    if(isEmpty()){
        return 0 ; 
    }
    var temp = root ; 
    root = root.next ; 

    return temp.data ; 
}

function peek(){
    if(isEmpty()){
        console.log("Stack is empty .") ; 
        return 0 ; 
    }else{
        return root.data ; 
    }
}

push(20) ; 
push(30) ;
push(40) ; 
console.log("Element current : ", peek()) 
console.log("Popped element : ", pop()) 
